File: sjui_tools/lib/core/image_accessibility.rb
# frozen_string_literal: true

module JsonUIShared
  # What a screen reader hears for each image of a layout — one rule for the
  # sjui and kjui codegen (the Dynamic runtimes of both libraries implement
  # the same rule, and all four run shared/core/image_accessibility_vectors
  # .json).
  #
  # `alt` (aliases `accessibilityLabel`, `contentDescription`) is the image's
  # spoken text: a strings.json key or literal, localized like `text`, or a
  # binding. An image is one of three roles:
  #
  #   label      — alt is a non-empty string: it is read.
  #   decorative — alt is "", or there is no alt and the image operates
  #                nothing: screen readers skip it (web alt="").
  #   control    — there is no alt and the image operates a control: it has a
  #                tap handler of its own, or it sits in the nearest tappable
  #                whose content names nothing (no text, no labelled image).
  #
  # A control image keeps what each platform read before alt existed (the id
  # on Android, the asset name on iOS), and the build names it (INFO). Hiding
  # it would take the control away from VoiceOver altogether: a tappable is a
  # `.onTapGesture`, not a Button, so with its only image hidden it has no
  # accessibility element left. That is the silence this rule chooses: a
  # control image reads an internal name until someone gives it an alt,
  # visibly (INFO) but without failing the build.
  module ImageAccessibility
    module_function

    # Image and its type aliases (component_metadata.json) plus NetworkImage.
    IMAGE_TYPES = %w[Image CircleImage CircleImageView ImageView Img NetworkImage].freeze

    # The canonical spelling first; the aliases are declared on `alt` in
    # attribute_definitions.json. A layout that `jui build` normalized
    # carries only `alt`.
    ALT_KEYS = %w[alt accessibilityLabel contentDescription].freeze

    # What a screen-reader user activates: a tap and a long press.
    TAP_KEYS = %w[onClick onclick onLongPress].freeze

    # Text that names a control when it sits inside it (string_manager_core's
    # STRING_PROPERTIES less `alt`, which is read off images below). On an
    # image, `hint` / `placeholder` name a placeholder IMAGE, not text.
    TEXT_KEYS = %w[text hint placeholder label prompt].freeze

    ROLE_KEY = '_imageRole'

    def image?(node)
      node.is_a?(Hash) && IMAGE_TYPES.include?(node['type'])
    end

    # The image's alt as written, or nil when it declares none. Spelled out
    # key by key, in ALT_KEYS order (the spec holds the two together), so
    # `jui conformance coverage` sees sjui and kjui read `alt` here.
    def alt(node)
      return node['alt'] if node.key?('alt')
      return node['accessibilityLabel'] if node.key?('accessibilityLabel')
      return node['contentDescription'] if node.key?('contentDescription')

      nil
    end

    def tappable?(node)
      node.is_a?(Hash) && TAP_KEYS.any? { |key| node.key?(key) }
    end

    def children(node)
      %w[child children].flat_map do |key|
        value = node[key]
        case value
        when Array then value.select { |c| c.is_a?(Hash) }
        when Hash then [value]
        else []
        end
      end
    end

    # True when something inside `node` (itself included) gives a control a
    # name: text on a non-image, or an image whose alt is not empty.
    def names_something?(node)
      if image?(node)
        value = alt(node)
        return value.is_a?(String) && !value.empty?
      end
      return true if TEXT_KEYS.any? { |key| node[key].is_a?(String) && !node[key].empty? }
      return true if node['items'].is_a?(Array) && node['items'].any? { |i| i.is_a?(String) && !i.empty? }

      children(node).any? { |c| names_something?(c) }
    end

    # The role of one image, given the tappables above it (nearest last).
    def role(node, tappable_ancestors = [])
      value = alt(node)
      return(value.to_s.empty? ? 'decorative' : 'label') unless value.nil?
      return 'control' if tappable?(node)

      nearest = tappable_ancestors.last
      return 'control' if nearest && !names_something?(nearest)

      'decorative'
    end

    # Writes ROLE_KEY on every image of an include-expanded tree and returns
    # one INFO per control image, in the LayoutValidator warning shape.
    def annotate!(root, source_path:)
      infos = []
      walk(root, []) do |node, tappables|
        node[ROLE_KEY] = role(node, tappables)
        next unless node[ROLE_KEY] == 'control'

        name = node['id'] || node['srcName'] || node['src'] || node['url'] || node['type']
        infos << {
          level: :info,
          message: "#{node['type']} '#{name}' operates a control and has no alt, so screen " \
                   'readers hear its id or asset name. Add alt: a strings.json key (or text) ' \
                   'that says what the control does.',
          location: source_path
        }
      end
      infos
    end

    def walk(node, tappables, &block)
      return unless node.is_a?(Hash)

      yield node, tappables if image?(node)
      inner = tappable?(node) ? tappables + [node] : tappables
      children(node).each { |c| walk(c, inner, &block) }
    end
  end
end

File: shared/core/image_accessibility.py
"""What a screen reader hears for each image of a layout — the rule of
shared/core/image_accessibility.rb, for the Python side (`jsonui-test
validate`). Five implementations share it: this one, the Ruby module the sjui
and kjui codegen use, and the KotlinJsonUI and SwiftJsonUI Dynamic runtimes.
All of them run shared/core/image_accessibility_vectors.json.

An image is one of three roles:

  label      — alt is a non-empty string: it is read.
  decorative — alt is "", or there is no alt and the image operates nothing.
               iOS hides it from VoiceOver, and the iOS driver then reads it as
               not visible even while it is on screen.
  control    — there is no alt and the image operates a control (a tap
               handler of its own, or it is the only thing naming the nearest
               tappable around it).

Loaded through `jui_cli.core.shared_core.load("image_accessibility")`; it
imports nothing of its own.
"""

#: Image, its type aliases (component_metadata.json) and NetworkImage.
IMAGE_TYPES = frozenset({"Image", "CircleImage", "CircleImageView", "ImageView", "Img", "NetworkImage"})

#: The canonical spelling first, then the aliases declared on `alt`.
ALT_KEYS = ("alt", "accessibilityLabel", "contentDescription")

#: What a screen-reader user activates, read as shared/core/tap_accessibility.rb
#: reads it (this module imports nothing, so the tap rule's predicate is copied
#: here, and the shared vectors hold the two together): a tap is a handler on
#: onClick / onclick, and a long press one on onLongPress.
TAP_KEYS = ("onClick", "onclick")
LONG_PRESS_KEY = "onLongPress"

#: Text that names a control it sits in (on an image, hint / placeholder name an image).
TEXT_KEYS = ("text", "hint", "placeholder", "label", "prompt")


def is_image(node) -> bool:
    return isinstance(node, dict) and node.get("type") in IMAGE_TYPES


def alt(node):
    """The image's alt as written, or None when it declares none (JSON null counts as none)."""
    for key in ALT_KEYS:
        if key in node:
            return node[key]
    return None


def names_a_method(value) -> bool:
    """A handler names a method that is not blank: a binding's inside
    (`@{onOpen}`), a bare selector, or each string of an `onclick` array.
    Blank is Unicode white space (`str.isspace`, a full-width space too)."""
    if not isinstance(value, str):
        return False
    inner = value[2:-1] if value.startswith("@{") and value.endswith("}") else value
    return inner.strip() != ""


def is_handler(value) -> bool:
    values = value if isinstance(value, list) else [value]
    return any(names_a_method(v) for v in values)


def is_tappable(node) -> bool:
    """Whether `node` operates something a screen-reader user can activate —
    a tap (a handler, `enabled` not false, `canTap` not false) or a long press
    (a handler, `enabled` not false) — as the tap rule judges it. It read the
    handler KEY before, so an empty, disabled or shut tap made an image a
    control. A bound gate still operates: it opens at run time."""
    if not isinstance(node, dict) or node.get("enabled") is False:
        return False
    if node.get("canTap") is not False and any(is_handler(node.get(key)) for key in TAP_KEYS):
        return True
    return is_handler(node.get(LONG_PRESS_KEY))


def children(node) -> list:
    out = []
    for key in ("child", "children"):
        value = node.get(key)
        if isinstance(value, list):
            out.extend(c for c in value if isinstance(c, dict))
        elif isinstance(value, dict):
            out.append(value)
    return out


def _non_empty_string(value) -> bool:
    return isinstance(value, str) and value != ""


def names_something(node) -> bool:
    """True when something inside `node` (itself included) names a control:
    text on a non-image, or an image whose alt is a non-empty string."""
    if is_image(node):
        return _non_empty_string(alt(node))
    if any(_non_empty_string(node.get(key)) for key in TEXT_KEYS):
        return True
    items = node.get("items")
    if isinstance(items, list) and any(_non_empty_string(i) for i in items):
        return True
    return any(names_something(c) for c in children(node))


def role(node, nearest_tappable=None) -> str:
    """The role of one image, given the nearest tappable around it (None when none)."""
    value = alt(node)
    if value is not None:
        return "decorative" if (value if isinstance(value, str) else str(value)) == "" else "label"
    if is_tappable(node):
        return "control"
    if nearest_tappable is not None and not names_something(nearest_tappable):
        return "control"
    return "decorative"


def roles(root) -> list:
    """(image node, role) for every image of a layout tree, in document order."""
    out: list = []

    def walk(node, nearest):
        if not isinstance(node, dict):
            return
        if is_image(node):
            out.append((node, role(node, nearest)))
        inner = node if is_tappable(node) else nearest
        for c in children(node):
            walk(c, inner)

    walk(root, None)
    return out
